storage/boltdb/client.go

```diff
@@ -50,136 +38,254 @@ func NewClient(logger *zap.Logger, path, bucket string) (*Client, error) {
+func (client *Client) ReverseList(first storage.Key, limit storage.Limit) (storage.Keys, error) {
+	return storage.ReverseListKeys(client, first, limit)
+}
+
+// Close closes a BoltDB client
+func (client *Client) Close() error {
+	return client.db.Close()
+}
+
+// GetAll finds all values for the provided keys up to 100 keys
+// if more keys are provided than the maximum an error will be returned.
+func (client *Client) GetAll(keys storage.Keys) (storage.Values, error) {
+	lk := len(keys)
+	if lk > maxKeyLookup {
+		return nil, Error.New(fmt.Sprintf("requested %d keys, maximum is %d", lk, maxKeyLookup))
+	}
+
+	vals := make(storage.Values, 0, lk)
+	err := client.view(func(bucket *bolt.Bucket) error {
+		for _, key := range keys {
+			val := bucket.Get([]byte(key))
+			vals = append(vals, storage.CloneValue(storage.Value(val)))
 		}
 		return nil
 	})
-	return paths, err
+
+	return vals, err
 }
 
-func firstOrLast(reverseList bool, cur *bolt.Cursor) func() ([]byte, []byte) {
-	if reverseList {
-		return cur.Last
-	}
-	return cur.First
+// Iterate iterates over collapsed items with prefix starting from first or the next key
+func (client *Client) Iterate(prefix, first storage.Key, delimiter byte, fn func(storage.Iterator) error) error {
+	return client.iterate(prefix, first, false, delimiter, fn)
 }
 
-func prevOrNext(reverseList bool, cur *bolt.Cursor) func() ([]byte, []byte) {
-	if reverseList {
-		return cur.Prev
-	}
-	return cur.Next
+// IterateAll iterates over all items with prefix starting from first or the next key
+func (client *Client) IterateAll(prefix, first storage.Key, fn func(storage.Iterator) error) error {
+	return client.iterate(prefix, first, true, '/', fn)
 }
 
-//ListV2 is the new definition and will replace `List` definition
-func (c *Client) ListV2(opts storage.ListOptions) (storage.Items, storage.More, error) {
-	//TODO write the implementation
-	panic("to do")
+// IterateReverse iterates over collapsed items with prefix starting from first or the prev key
+func (client *Client) IterateReverse(prefix, first storage.Key, delimiter byte, fn func(storage.Iterator) error) error {
+	return client.iterateReverse(prefix, first, false, delimiter, fn)
 }
 
-// Delete deletes a key/value pair from boltdb, for a given the key
-func (c *Client) Delete(pathKey storage.Key) error {
-	c.logger.Debug("entering bolt delete: " + string(pathKey))
-	return c.db.Update(func(tx *bolt.Tx) error {
-		return tx.Bucket(c.Bucket).Delete(pathKey)
-	})
+// IterateReverseAll iterates over all items with prefix starting from first or the prev key
+func (client *Client) IterateReverseAll(prefix, first storage.Key, fn func(storage.Iterator) error) error {
+	return client.iterateReverse(prefix, first, true, '/', fn)
 }
 
-// Close closes a BoltDB client
-func (c *Client) Close() error {
-	return c.db.Close()
+func (client *Client) iterate(prefix, first storage.Key, recurse bool, delimiter byte, fn func(storage.Iterator) error) error {
```

Comment:
We should choose between `recurse` and `delimiter`, but not both. If the delimiter is empty then the iteration is recursive. If the delimiter is non-empty then the iteration is non-recursive. It does not make sense to have a recursive iteration with a delimiter.